### PR TITLE
New lint `format_add_strings`

### LINKED ISSUE #6261: Warn on push_str(&format!())
### What it does

Emits a warning on code like ```rust
some_string.push_str(&format!(...));
// or
some_string += &format!(...);
```
and suggests rewriting it to ```rust
write!(&mut some_string, ...);
```

### Categories
- Kind: `Perf`

*What is the advantage of the recommended code over the original code*
It avoids one extra heap allocation by writing directly to the existing String instead of creating a new temporary String.

### Drawbacks
The user has to `use std::io::Write` for the `write!` macro to work (which I don't get tbh, why can't `write!` just import the trait on its own...)

### Example
```rust
let mut string = String::new();
for foo in foo_list {
    string += &format!("{:?}", foo);
}
```

Could be written as:

```rust
use std::io::Write;

let mut string = String::new();
for foo in foo_list {
    write!(&mut string, "{:?}", foo);
}
```

### PR DESCRIPTION
**Pull request by pitaj**
Closes #6261 

changelog: Added [`format_add_string`]: recommend using `write!` instead of appending the result of  `format!`

### CODE REVIEW

File: clippy_lints/src/format_push_string.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    ///
+    /// ### Example
+    /// ```rust
+    /// let mut s = String::new();
+    /// s += &format!("0x{:X}", 1024);
+    /// s.push_str(&format!("0x{:X}", 1024));
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::fmt::Write;
```
Comment (llogiq) [inline clippy_lints/src/format_push_string.rs:25]: Should we suggest `use std::fmt::Write as _;` here to avoid putting the item name in scope?

**Review by llogiq — APPROVED** [review]
This looks good to me. I have but one small suggestion and you will need to rebase the changes. Once that's done, I'll be happy to merge.

**Comment by pitaj** [comment]
Changed to recommend `as _` for import and rebased.

**Comment by llogiq** [comment]
Thanks! It will need another rebase, then it'll be ready to merge. Feel free to r+ it yourself

@bors delegate+

**Comment by pitaj** [comment]
@llogiq why does it need another rebase? There are no conflicts.

**Comment by xFrednet** [comment]
Maybe he missed your push? I'd say we let bors decide if this is ready or not. I don't want to steal your honer to use bors. You can merge this by commenting `@bors r+ llogiq`

**Comment by llogiq** [comment]
In that case @bors r+

**Pull request merged**